currency/storage_test.go

```diff
@@ -30,219 +30,92 @@ func TestMain(m *testing.M) {
-	err = newStorage.RunUpdater(BotOverrides{ExchangeRates: true, OpenExchangeRates: true}, &mainConfig, tempDir)
-	if !errors.Is(err, nil) {
-		t.Fatalf("received: '%v' but expected: '%v'", err, nil)
-	}
-
-	if mainConfig.ForexProviders[0].Enabled {
-		t.Fatal("old default provider should not be enabled due to unset keys")
-	}
-
-	if mainConfig.ForexProviders[0].PrimaryProvider {
-		t.Fatal("old default provider should not be a primary provider anymore")
-	}
-
-	if mainConfig.ForexProviders[1].Enabled {
-		t.Fatal("open exchange rates provider with keys unset should not be enabled")
-	}
-
-	if mainConfig.ForexProviders[1].PrimaryProvider {
-		t.Fatal("open exchange rates provider with keys unset should not be set as primary provider")
-	}
-
-	if !mainConfig.ForexProviders[2].Enabled {
-		t.Fatal("actual default provider should not be disabled")
-	}
+	assert.NoError(t, err, "Shutdown should not error evne though it silently aborted the RunUpdater early")
+
+	// Exchanges which reject a bad APIKey
+	for _, n := range []string{"Fixer", "CurrencyConverter", "CurrencyLayer", "ExchangeRates"} {
+		c.ForexProviders = AllFXSettings{{Name: n, Enabled: true, APIKey: ""}}
+		err = newStorage.RunUpdater(overideForProvider(n), c, tempDir)
+		assert.NoErrorf(t, err, "%s should not error and silently exit without running with no api keys", n)
+		assert.Falsef(t, c.ForexProviders[0].Enabled, "%s should not be marked enabled with no api keys", n)
+		assert.Nil(t, newStorage.fiatExchangeMarkets, "Forex should not be enabled with no providers")
+		c.ForexProviders = AllFXSettings{{Name: n, Enabled: true, APIKey: "sudo shazam!"}}
+		err = newStorage.RunUpdater(overideForProvider(n), c, tempDir)
+		assert.Errorf(t, err, "%s should throw some provider originating error with a (hopefully) invalid api key", n)
+		assert.Truef(t, c.ForexProviders[0].Enabled, "%s should still be enabled after being chosen but failing", n)
+		assert.Nil(t, newStorage.fiatExchangeMarkets, "Forex should not be enabled when provider errored during startup")
+		err = newStorage.Shutdown()
+		assert.NoError(t, err, "Shutdown should not error")
+	}
+
+	// Exchanges which do not error with a bad APIKey on startup
+	for _, n := range []string{"OpenExchangeRates", "ExchangeRateHost"} {
+		c.ForexProviders = AllFXSettings{{Name: n, Enabled: true, APIKey: ""}}
+		err = newStorage.RunUpdater(overideForProvider(n), c, tempDir)
+		assert.NoErrorf(t, err, "%s should not error and silently exit without running with no api keys", n)
+		assert.Nil(t, newStorage.fiatExchangeMarkets, "Forex should not be enabled with no providers")
+		c.ForexProviders = AllFXSettings{{Name: n, Enabled: true, APIKey: "sudo shazam!"}}
+		err = newStorage.RunUpdater(overideForProvider(n), c, tempDir)
+		assert.NoError(t, err, "%s should not error on Setup with a bad apikey", n)
+		assert.NotNil(t, newStorage.fiatExchangeMarkets, "Forex should not be enabled with no providers")
```

Comment:
<img width="1536" alt="image" src="https://github.com/thrasher-corp/gocryptotrader/assets/9261323/3eda9ec4-c57f-4a9f-b20b-33cf16bb06ff">
Inaccurate commentary? Or more specific check required?

Reply:
Inaccurate Human. Bad Human.
Thanks for spotting.

Fixed 8085342